### ISSUE TITLE
Running spider via CrawlerRunner from script gives Error: ReactorNotRestartable

### ISSUE THREAD

**Issue by ArionMiles** [scrapy/scrapy#2941]
Ok so I've given it a lot of try, asked most pythonistas I know, even on stackoverflow here: https://stackoverflow.com/questions/46346863/twisted-reactor-not-restarting-in-scrapy

But I cannot for the life of me figure this out. I'm trying to execute the spider via a telegram bot, using the python-telegram-bot API wrapper.

I'm running Python 2.7 on Windows 10.

This is my code:
```
from twisted.internet import reactor
from scrapy import cmdline
from telegram.ext import Updater, CommandHandler, MessageHandler, Filters, RegexHandler
import logging
import os
import ConfigParser
import json
import textwrap
from MIS.spiders.moodle_spider import MySpider
from scrapy.utils.project import get_project_settings
from scrapy.crawler import CrawlerRunner, CrawlerProcess
from scrapy.utils.log import configure_logging


# Read settings from config file
config = ConfigParser.RawConfigParser()
config.read('./spiders/creds.ini')
TOKEN = config.get('BOT', 'TOKEN')
#APP_NAME = config.get('BOT', 'APP_NAME')
#PORT = int(os.environ.get('PORT', '5000'))
updater = Updater(TOKEN)

# Setting Webhook
#updater.start_webhook(listen="0.0.0.0",
#                      port=PORT,
#                      url_path=TOKEN)
#updater.bot.setWebhook(APP_NAME + TOKEN)

logging.basicConfig(format='%(asctime)s -# %(name)s - %(levelname)s - %(message)s',level=logging.INFO)

dispatcher = updater.dispatcher

def doesntRun(bot, update):
    configure_logging({'LOG_FORMAT': '%(levelname)s: %(message)s'})
    runner = CrawlerRunner({
        'FEED_FORMAT' : 'json',
        'FEED_URI' : 'output.json'
        })

    d = runner.crawl(MySpider)
    d.addBoth(lambda _: reactor.stop()) # this line is supposed to restart the reactor, right?
    reactor.run(installSignalHandlers=0) # the script will block here until the crawling is finished

    with open("./output.json", 'r') as file:
        contents = file.read()
        a_r = json.loads(contents)
        AM = a_r[0]['AM']
        ...
        ...

        message_template = textwrap.dedent("""
                AM: {AM}
                ...
                """)
        messageContent = message_template.format(AM=AM, ...)
        bot.sendMessage(chat_id=update.message.chat_id, text=messageContent)


# Handlers
test_handler = CommandHandler('doesntRun', doesntRun)

# Dispatchers
dispatcher.add_handler(test_handler)

updater.start_polling()
updater.idle()
```

Please provide insight on how I can restart the reactor, this is bugging me from a couple of days.

**Comment by IAlwaysBeCoding** [comment]
`CrawlerRunner` runs a crawler but does not take care of the [install_shutdown_handler,configure_logging,log_scrapy_info] (https://github.com/scrapy/scrapy/crawler.py#L247-L249). Like the docs say that `CrawlRunner` should only be used if you are using it from a reactor, but it won't be able to run twice because it is missing the code found inside `start()` from the `CrawlerProcess` code. 

If you look very carefully at the code, you will see that `CrawlerProcess` is ran by calling the `start()` method. In your code, which you just copy and pasted without understanding what it does, you import the `CrawlerProcess` but you fail to use it. 

Basically, the way it works is that you start a new `CrawlerProcess` instance, then you pass a spider to class by calling `crawl()`(which you have done correctly with the `CrawlRunner`). After that, you call the `start()` method on the `CrawlerProcess` instance.

This makes it hard to implement with what you are doing ,unless you implement the `start()` logic from `CrawlerProcess` inside your script.

Otherwise, `CrawlRunner` is only meant to be ran ONCE, because it always tries to restart the reactor.

**Comment by ArionMiles** [comment]
I actually tried using the `CrawlerProcess` earlier, which is why you see it's imported, I had commented out the `CrawlerProcess` code and deleted it to make this MRE.

Even when I use `CrawlerProcess`, I get `signal only works in main thread` error, which I can avoid in `CrawlerRunner` since I can easily pass the `installSignalHandlers=0` argument in the `reactor.run()` but I don't know how to do it with `CrawlerProcess`. If you could tell me how, I'd rather use CrawlerProcess as I originally intended do, than deal with twisted reactors.

**Comment by IAlwaysBeCoding** [comment]
I guess you didn't really understand. 

Imagine it like this:

Telegram bot runs a machine that has its own functionality with a lot of other stuff running in the background like webhooks, updater, etc.  and can run a CALLBACK after each handler is executed, in your case `CommandHandler`. The only way to hook anything is via this CALLBACK. 

Scrapy runs another machine using the Twisted reactor as the asynchronous loop, and uses the Spider class to yield `Request` , `dict` and/or `Item` instances. Now, on the middlewares, pipelines, and extensions it can yield these things called `deferreds` which are functions with callbacks, and error callbacks.

You are trying to mix 2 different machines that only talk to each other through callbacks, and run their own asynchronous loop. If you want to get this to work, you need to get these 2 machines to run on the SAME asynchronous loop. 

You need to pass the functionality of one machine onto the other , and attach the callbacks that they run, not just the CommandHandler callback, but everything else it does on the background.

What you are trying to do is time consuming and hard, it is not a simple add a callback. The documents at Scrapy only say how to run your own Scrapy spiders if you are using other programs that run Twisted. Your telegram library does not run a Twisted reactor, so it is a different machine that you need to implement on top of Scrapy.

Does it make sense?

**Comment by ArionMiles** [comment]
Ok, so what's the best way to go about doing this then?

**Comment by ArionMiles** [comment]
Okay, I finally solved my problem.

the Python-telegram-bot API wrapper offers [an easy way to restart the bot.](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Code-snippets#simple-way-of-restarting-the-bot)

I simply put the lines:
```
time.sleep(0.2)
os.execl(sys.executable, sys.executable, *sys.argv)
```
at the end of the `doesntRun()` function. Now whenever I call the function via bot, it scrapes the page, stores the results, forwards the result, then restarts itself. Doing so allows me to execute the spider any number of times I want.

**Comment by WalkOnMars** [comment]
> Okay, I finally solved my problem.
> 
> the Python-telegram-bot API wrapper offers [an easy way to restart the bot.](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Code-snippets#simple-way-of-restarting-the-bot)
> 
> I simply put the lines:
> 
> ```
> time.sleep(0.2)
> os.execl(sys.executable, sys.executable, *sys.argv)
> ```
> 
> at the end of the `doesntRun()` function. Now whenever I call the function via bot, it scrapes the page, stores the results, forwards the result, then restarts itself. Doing so allows me to execute the spider any number of times I want.

amazing solution!!! thanks a lot~~

**Comment by ArionMiles** [comment]
@WalkOnMars I later found out this method is not recommended. There's a better way to do this, which is to spawn a separate process which executes the spider when your bot function is executed.

This SO answer has a code example with it:
https://stackoverflow.com/a/43661172/5129096

And this is how I've used the above answer in my own code:
https://github.com/ArionMiles/MIS-Bot/blob/master/mis_bot/scraper/spiders/attendance_spider.py#L104

**Comment by chipzzz** [comment]
@ArionMiles , it's a decent fix,  I'm running in a container so multiple process not recommended , so am using the kill process method that you provided for now.

I am surprised that this doesn't work out of the box, which is what they recommend, but I would still get the same error

```from twisted.internet import reactor
import scrapy
from scrapy.crawler import CrawlerRunner
from scrapy.utils.log import configure_logging

class MySpider(scrapy.Spider):
    # Your spider definition
    ...

configure_logging({'LOG_FORMAT': '%(levelname)s: %(message)s'})
runner = CrawlerRunner()

d = runner.crawl(MySpider)
d.addBoth(lambda _: reactor.stop())
reactor.run() # the script will block here until the crawling is finished
```
